akka-stream/src/main/scala/akka/stream/impl/ActorRefSourceActor.scala

```diff
@@ -46,7 +46,7 @@ import akka.stream.ActorMaterializerSettings
     .orElse(receiveElem)
 
   def receiveComplete: Receive = completionMatcher.andThen { _ ⇒
-    if (bufferSize == 0 || buffer.isEmpty) context.stop(self) // will complete the stream successfully
+    if (bufferSize == 0 || buffer.isEmpty) onCompleteThenStop() // will complete the stream successfully
```

Comment:
Can you explain why `ActorPublisher.aroundPostStop` did not take care of this?

Reply:
This is my first dive into the internals of akka so I'm not sure I can answer this one. Nearest I can tell the only thing `onCompleteThenStop` does particularly differently is it handles the `PreSubscriber` state as well as the `Active` state; this is something @patriknw was alluding to above I think, but without understanding the wider context of this subscription model I can't comment on why it might be in a pre-subscriber state at the point in which I'm sending the Success in my test; indeed I would have expected these internal subscriptions to all be established once the stream has started.

Reply:
I've just been playing around with this in ammonite and it completes ok when I play around actively (with several seconds until I try to complete) so presumably the answer is that it's a race and the stream will fail to complete if it's still in the process of setting up the subscriber, which may still be incomplete when the ActorRef materialises. My test consistently fails without the fix, but that's simply probably because sending a couple of messages right after the actor materializes is consistently faster than establishing that subscription.

Reply:
It's a race condition when sending the PoisonPill before everything is up and running completely